### PR TITLE
Superclass relations

### PR DESCRIPTION
**Pull request by b-mehta**
Add superclasses Data.Functor => Control.Functor and similarly for Applicative; along with default definitions to conveniently define such instances.

### CODE REVIEW

File: src/Control/Monad/Linear.hs [hunk 1]
```diff
@@ -34,21 +37,27 @@ import Prelude (String)
...
+class (Data.Applicative f, Functor f) => Applicative f where
   {-# MINIMAL pure, ((<*>) | liftA2) #-}
   pure :: a ->. f a
   (<*>) :: f (a ->. b) ->. f a ->. f b
   (<*>) = liftA2 id
   liftA2 :: (a ->. b ->. c) ->. f a ->. f b ->. f c
   liftA2 f x y = f <$> x <*> y
 
+dataPureDefault :: Applicative f => a -> f a
+dataPureDefault x = pure x
```
Comment (aspiwack) [inline src/Control/Monad/Linear.hs]: Maybe we could also straight up give `DerivingVia` combinator for these.

Like, define a data type

```haskell
newtype Forget f a = Forget (f a)
```

Then prove

```haskell
instance Functor f => Data.Forget (Control f)
instance Applicative f => Data.Applicative (Control f)
```

Then we can use it in `deriving` clauses.

Comment (b-mehta) [reply src/Control/Monad/Linear.hs]: Do you mean 
```haskell
instance Functor f => Data.Functor (Forget f)
instance Applicative f => Data.Applicative (Forget f)
```

Comment (aspiwack) [reply src/Control/Monad/Linear.hs]: Yes. Let's blame it on the heat.

Comment (b-mehta) [reply src/Control/Monad/Linear.hs]: I've added this, with a different name - since the name is exported I felt something more evocative would be nicer.

Comment (b-mehta) [reply src/Control/Monad/Linear.hs]: I wonder if it is worth adding 
```
Linear.coerce :: Coercible a b => a ->. b
```
also.

File: src/Control/Monad/Linear.hs [hunk 2]
```diff
@@ -79,3 +80,13 @@ join action = action >>= id
 -- | Convenience operator to define Applicative instances in terms of Monad
 ap :: Monad m => m (a ->. b) ->. m a ->. m b
 ap f x = f >>= (\f' -> fmap f' x)
+
+-- | DerivingVia combinators for Data.XXX in terms of Control.XXX
+newtype DataFromControl f a = DataFromControl (f a)
```
Comment (aspiwack) [inline src/Control/Monad/Linear.hs]: It's a bit wordy as it's supposed to be imported qualified. So how about `Control.Data` instead?

Comment (b-mehta) [reply src/Control/Monad/Linear.hs]: Ah good point!
